src/Form/FormAutosuggest.jsx

```diff
@@ -14,292 +15,358 @@ import Spinner from '../Spinner';
+        });
+      });
 
-    setIsMenuClosed(true);
-  };
+      if (strToFind.length > 0) {
+        childrenOpt = childrenOpt
+          .filter((opt) => (opt.props.children.toLowerCase().includes(strToFind.toLowerCase())));
+      }
 
-  const handleDocumentClick = (e) => {
-    if (parentRef.current && !parentRef.current.contains(e.target) && isActive) {
-      leaveControl();
+      return childrenOpt;
     }
-  };
 
-  const keyDownHandler = e => {
-    if (e.key === 'Escape' && isActive) {
-      e.preventDefault();
+    const expandDropdown = () => {
+      setDropdownItems(getItems(displayValue));
+      setIsValid(true);
+      setErrorMessage('');
+      setDropdownExpanded(true);
+    };
 
-      if (formControlRef) {
-        formControlRef.current.focus();
+    const toggleDropdown = () => {
+      if (dropdownExpanded) {
+        collapseDropdown();
+      } else {
+        expandDropdown();
       }
+    };
 
-      setState(prevState => ({
-        ...prevState,
-        dropDownItems: [],
-      }));
+    const iconToggle = (
+      <IconButton
+        className="pgn__form-autosuggest__icon-button"
+        data-testid="autosuggest-iconbutton"
+        tabIndex="-1"
+        src={dropdownExpanded ? KeyboardArrowUp : KeyboardArrowDown}
+        iconAs={Icon}
+        size="sm"
+        variant="secondary"
+        alt={dropdownExpanded
+          ? intl.formatMessage(messages.iconButtonClosed)
+          : intl.formatMessage(messages.iconButtonOpened)}
+        onClick={toggleDropdown}
+      />
+    );
+
+    const enterControl = () => {
+      setIsActive(true);
+    };
 
-      setIsMenuClosed(true);
-    }
-    if (e.key === 'Tab' && isActive) {
-      leaveControl();
-    }
-  };
+    const updateErrorStateAndErrorMessage = () => {
+      if (customError) {
+        setIsValid(false);
+        setErrorMessage(customErrorMessageText);
+        return;
+      }
 
-  useEffect(() => {
-    document.addEventListener('keydown', keyDownHandler);
-    document.addEventListener('click', handleDocumentClick, true);
+      if (valueRequired && !hasValue) {
+        setIsValid(false);
+        setErrorMessage(valueRequiredErrorMessageText);
+        return;
+      }
 
-    return () => {
-      document.removeEventListener('click', handleDocumentClick, true);
-      document.removeEventListener('keydown', keyDownHandler);
+      if (hasValue && selectionRequired && !hasSelection) {
+        setIsValid(false);
+        setErrorMessage(selectionRequiredErrorMessageText);
+        return;
+      }
+
+      setIsValid(true);
+      setErrorMessage('');
     };
-  });
-
-  useEffect(() => {
-    if (value || value === '') {
-      setState(prevState => ({
-        ...prevState,
-        displayValue: value,
-      }));
-    }
-  }, [value]);
 
-  const setDisplayValue = (itemValue) => {
-    const optValue = [];
+    useImperativeHandle(ref, () => ({
+      updateErrorStateAndErrorMessage,
+    }));
```

Comment:
[nit/suggestion] May be a good idea to include a comment here or elsewhere documenting the use of `useImperativeHandle` to only include `updateErrorStateAndErrorMessage` on its exposed methods. That way, future contributions will have some guidance on how to keep this maintained such that methods intended to be exposed are intentionally included here? 🤔

Reply:
added a comment here https://github.com/openedx/paragon/compare/f8a4e5ee118199936c3ff6653e3dfb389d12fffa..9310aa04d988eba1ee524cf920e5ad54dd39590f#diff-96e13c1a773ec8d218ea5667d3dec6c8f6eeadcb00c81584f50d6c3d4d52135fR170-R171

If you have another suggestion for a comment I'm more than happy to change it!